logical_switch_port.go

```diff
@@ -235,20 +249,30 @@ func (odbi *ovndb) lspGetDynamicAddressesImp(lsp string) (string, error) {
 }
 
 func (odbi *ovndb) lspSetExternalIdsImp(lsp string, external_ids map[string]string) (*OvnCommand, error) {
+	if external_ids == nil {
+		return nil, ErrorOption
+	}
+
 	if len(lsp) == 0 {
-		return nil, fmt.Errorf("LSP name cannot be empty while setting external ids")
+		return nil, fmt.Errorf("LSP name cannot be empty while setting external_ids")
 	}
 
-	mutatemap, _ := libovsdb.NewOvsMap(external_ids)
-	mutation := libovsdb.NewMutation("external_ids", opInsert, mutatemap)
+	mutatemap, err := libovsdb.NewOvsMap(external_ids)
+	if err != nil {
+		return nil, err
+	}
+
+	row := make(OVNRow)
+	row["external_ids"] = mutatemap
+
 	condition := libovsdb.NewCondition("name", "==", lsp)
 
 	// simple mutate operation
 	mutateOp := libovsdb.Operation{
-		Op:        opMutate,
-		Table:     tableLogicalSwitchPort,
-		Mutations: []interface{}{mutation},
-		Where:     []interface{}{condition},
+		Op:    opUpdate,
```

Comment:
This way, it changes the API behavior. Originally, the API only adds key-values on top of existing ones. Now it would completely replace the old ones. E.g. if there are k1:v1, k2:v2 existed, but the API caller just set k3:v3, then in the end k1 and k2 will both disappear. I am not sure if this is the desired behavior of all existing interface use cases. If using update, it is better to:

1) read old key-values from cache

2) use "wait" operation to make sure it is not changed by anyone else in parallel

3) add the new k-v to existing map and update

You can check ovn-nbctl -v set ... external-ids:xxx=yyy to see the transaction details.

Alternatively, we can use mutation, but delete the old entries if existed.

Reply:
Yes, @hzhou8. But the original implementation was not correct. In particular, it will barf if the argument map contained existing keys. I agree this is a behavior change. But the Set API should clobber the existing map entries IMO. We can provide a separate AddOptions/AddExternalIds API to do the diff (i.e. the correct implementation of existing API logic).

Reply:
I don't know how many people were using the existing Set API to semantically mean that this is an additive change.

Reply:
@abhat make sense. The function name "set" suggests a full update. Not sure about impact for existing users. Maybe @vtolstov @softwarejl @noah8713 can comment. If there is no impact for this behavior change for your current use cases, I'd support change it to completely replace the whole column with this API. It is similar for lsp_set_options API.

Reply:
Agree with the update part. However, better confirm the current UT works with mutateOp and then add a new patch to cover add/set operations both. However, better add addextid support too so that for both set/add,  fetching from cache odbi.cache[][uuid].Fields["external_ids"].(libovsdb.OvsMap).GoMap is needed  and then update the k:v that needs to be set/add/ . I did confirm with nbctl command it seems idl takes all keys even if one key is being set:
existing external_ids: {aa=yyy, abc="456", xxx="122"}
Run to update xxx=123:. ovn-nbctl -v set logical_switch_port sw0-port1 external-ids:xxx=123
verbose-data: "external_ids":["map",[["aa","yyy"],["abc","456"],["xxx","123"]]]}}
For go client, we might need to check if we need to fetch from cache and explicitely update the key.

Reply:
I can implement the`AddExternalIds`/`AddOptions" API in a separate PR since that doesn't have any bearing with the changes here.

To summarize: 

SetOptions/SetExternalIds will do a full-update of the columns in question.

AddOptions/AddExternalIds will do a logical union of keys including mutating the intersecting keys' values.

Assuming we are in agreement here, I have updated the PR and fixed the failing UTs. I have also addressed @noah8713's previous comments. PTAL. I would like for this PR to get merged including the fix to not recursively call clearConnection in the libovsdb repo: https://github.com/eBay/libovsdb/pull/18.